test/persistence-hooks.suite.js

```diff
@@ -402,6 +449,112 @@ module.exports = function(dataSource, should) {
+        TestModel.observe('persist', pushContextAndNext());
+
+        TestModel.findOrCreate(
+          { where: { name: 'new-record' } },
+          { name: 'new-record' },
+          function(err, record, created) {
+            if (err) return done(err);
+
+            // Optimized connector context, does NOT have `currentInstance`
+            // whereas, unoptimized connector has it in the context
+            if (dataSource.connector.findOrCreate) {
+              observedContexts.should.eql(aTestModelCtx({
+                data: {
+                  id: record.id,
+                  name: 'new-record'
+                }
+              }));
+            } else {
+              observedContexts.should.eql(aTestModelCtx({
+                data: {
+                  id: record.id,
+                  name: 'new-record'
+                },
+                currentInstance: { id: record.id, name: record.name, extra: null }
+              }));
+            }
+            done();
+          });
+      });
+
+      if (dataSource.connector.findOrCreate) {
+        it('applies updates from `persist` hook when record found', function(done) {
+          TestModel.observe('persist', pushContextAndNext(function(ctx){
+            ctx.data.extra = 'hook data';
+          }));
+
+          TestModel.findOrCreate(
+            { where: { name: existingInstance.name } },
+            { name: existingInstance.name },
+            function(err, instance) {
+              if (err) return done(err);
+
+              // instance returned by `findOrCreate` context does not
+              // have the values updated from `persist` hook
+              instance.should.not.have.property('extra', 'hook data');
+
+              // Query the database. Here, since record already exists 
+              // `findOrCreate`, does not update database for
+              // updates from `persist` hook
+              TestModel.findById(existingInstance.id, function(err, dbInstance) {
+                if (err) return done(err);
+                dbInstance.toObject(true).should.eql({
+                  id: existingInstance.id,
+                  name: existingInstance.name,
+                  extra: undefined
+                });
+              });
+
+              done();
+            });
+        });
+      }
+
+      it('applies updates from `persist` hook - on findOrCreate when not found', function(done) {
+        TestModel.observe('persist', pushContextAndNext(function(ctx){
+          ctx.data.extra = 'hook data';
+        }));
+
+        TestModel.findOrCreate(
+          { where: { name: 'new-record' } },
+          { name: 'new-record' },
+          function(err, instance) {
+            if (err) return done(err);
+
+            if (dataSource.connector.findOrCreate) {
+              instance.should.have.property('extra', 'hook data');
+            } else {
+              instance.should.not.have.property('extra', 'hook data');
+            }
```

Comment:
Could you please add a comment explaining why different connectors behave differently?

Also please add a check that the record stored in the database always contains updated data.